pymongo/mongo_client.py

```diff
@@ -750,9 +767,14 @@ def target():
         self._encrypter = None
         if self.__options.auto_encryption_opts:
             from pymongo.encryption import _Encrypter
-            self._encrypter = _Encrypter.create(
+            self._encrypter = _Encrypter(
                 self, self.__options.auto_encryption_opts)
 
+    def _duplicate(self, **kwargs):
+        args = self.__init_kwargs.copy()
```

Comment:
Does this need to be a deepcopy?

Reply:
It can't hurt. Changed to deepcopy.

Reply:
Whoops. Looks like I led you astray. It did hurt:
```
 [2021/02/20 00:42:51.221] ERROR: test_case_2 (test_encryption.TestDeadlockProse)
 [2021/02/20 00:42:51.221] ----------------------------------------------------------------------
 [2021/02/20 00:42:51.221] Traceback (most recent call last):
 [2021/02/20 00:42:51.221]   File "/data/mci/ba1cc38b238f9453250d181b3de486eb/src/test/test_encryption.py", line 1441, in test_case_2
 [2021/02/20 00:42:51.221]     self._run_test(max_pool_size=1,
 [2021/02/20 00:42:51.221]   File "/data/mci/ba1cc38b238f9453250d181b3de486eb/src/test/test_encryption.py", line 1399, in _run_test
 [2021/02/20 00:42:51.221]     client_encrypted = rs_or_single_client(
 [2021/02/20 00:42:51.221]   File "/data/mci/ba1cc38b238f9453250d181b3de486eb/src/test/utils.py", line 532, in rs_or_single_client
 [2021/02/20 00:42:51.221]     return _mongo_client(h, p, **kwargs)
 [2021/02/20 00:42:51.221]   File "/data/mci/ba1cc38b238f9453250d181b3de486eb/src/test/utils.py", line 492, in _mongo_client
 [2021/02/20 00:42:51.221]     client = MongoClient(_connection_string(host, authenticate), port,
 [2021/02/20 00:42:51.221]   File "/data/mci/ba1cc38b238f9453250d181b3de486eb/src/pymongo/mongo_client.py", line 771, in __init__
 [2021/02/20 00:42:51.221]     self._encrypter = _Encrypter(
 [2021/02/20 00:42:51.221]   File "/data/mci/ba1cc38b238f9453250d181b3de486eb/src/pymongo/encryption.py", line 288, in __init__
 [2021/02/20 00:42:51.221]     metadata_client = _get_internal_client(self, client)
 [2021/02/20 00:42:51.221]   File "/data/mci/ba1cc38b238f9453250d181b3de486eb/src/pymongo/encryption.py", line 275, in _get_internal_client
 [2021/02/20 00:42:51.221]     internal_client = mongo_client._duplicate(
 [2021/02/20 00:42:51.221]   File "/data/mci/ba1cc38b238f9453250d181b3de486eb/src/pymongo/mongo_client.py", line 775, in _duplicate
 [2021/02/20 00:42:51.221]     args = copy.deepcopy(self.__init_kwargs)
 [2021/02/20 00:42:51.221]   File "/opt/python/3.9/lib/python3.9/copy.py", line 146, in deepcopy
 [2021/02/20 00:42:51.221]     y = copier(x, memo)
 [2021/02/20 00:42:51.221]   File "/opt/python/3.9/lib/python3.9/copy.py", line 230, in _deepcopy_dict
 [2021/02/20 00:42:51.221]     y[deepcopy(key, memo)] = deepcopy(value, memo)
 [2021/02/20 00:42:51.221]   File "/opt/python/3.9/lib/python3.9/copy.py", line 172, in deepcopy
 [2021/02/20 00:42:51.222]     y = _reconstruct(x, memo, *rv)
 [2021/02/20 00:42:51.222]   File "/opt/python/3.9/lib/python3.9/copy.py", line 270, in _reconstruct
 [2021/02/20 00:42:51.222]     state = deepcopy(state, memo)
 [2021/02/20 00:42:51.222]   File "/opt/python/3.9/lib/python3.9/copy.py", line 146, in deepcopy
 [2021/02/20 00:42:51.222]     y = copier(x, memo)
 [2021/02/20 00:42:51.222]   File "/opt/python/3.9/lib/python3.9/copy.py", line 230, in _deepcopy_dict
 [2021/02/20 00:42:51.222]     y[deepcopy(key, memo)] = deepcopy(value, memo)
 [2021/02/20 00:42:51.222]   File "/opt/python/3.9/lib/python3.9/copy.py", line 172, in deepcopy
 [2021/02/20 00:42:51.222]     y = _reconstruct(x, memo, *rv)
 [2021/02/20 00:42:51.222]   File "/opt/python/3.9/lib/python3.9/copy.py", line 270, in _reconstruct
 [2021/02/20 00:42:51.222]     state = deepcopy(state, memo)
 [2021/02/20 00:42:51.222]   File "/opt/python/3.9/lib/python3.9/copy.py", line 146, in deepcopy
 [2021/02/20 00:42:51.222]     y = copier(x, memo)
 [2021/02/20 00:42:51.222]   File "/opt/python/3.9/lib/python3.9/copy.py", line 230, in _deepcopy_dict
 [2021/02/20 00:42:51.222]     y[deepcopy(key, memo)] = deepcopy(value, memo)
 [2021/02/20 00:42:51.222]   File "/opt/python/3.9/lib/python3.9/copy.py", line 161, in deepcopy
 [2021/02/20 00:42:51.222]     rv = reductor(4)
 [2021/02/20 00:42:51.222] TypeError: cannot pickle '_thread.lock' object
 ```

Reply:
I can't really figure out where that is coming from. Reverted to a shallow copy. 

Now that I think about it, we don't really _need_ a shallow copy. Only top-level references will be mutated by the `dict.update` call and a shallow copy will be enough to prevent mutating `__init_kwargs`. In any case, this is (as of now) a private method and only used in one place so we can be fairly certain the shallow copy doesn't hurt us at the moment.